### ISSUE TITLE
Babel 6.0

### ISSUE THREAD

**Issue by stefanpenner** [ember-cli/ember-cli#5015]
Would like to cut over as soon as possible

```
npm WARN deprecated babel-core@5.8.33: Babel 5 is no longer being maintained. Upgrade to Babel 6.
```
- [x] update broccoli-babel-transpiler (https://github.com/babel/broccoli-babel-transpiler/issues/62)
- [x] update babel-plugin-feature-flags (https://github.com/ember-cli/babel-plugin-feature-flags/pull/6)
- [x] update babel-plugin-filter-imports (https://github.com/ember-cli/babel-plugin-filter-imports/pull/6)
- [x] update babel-plugin-htmlbars-inline-precompile (https://github.com/pangratz/babel-plugin-htmlbars-inline-precompile/pull/8)
- [ ] update ember-cli-htmlbars-inline-precompile ([ember-cli/ember-cli-htmlbars-inline-precompile](https://github.com/ember-cli/ember-cli-htmlbars-inline-precompile))
- [ ] update ember-cli-babel (https://github.com/babel/ember-cli-babel/issues/62)
- [ ] update emberjs-build (https://github.com/emberjs/emberjs-build/pull/165)

---

- [x] Migrate `ember-cli-babel` specific settings out of `babel` config and into its own `ember-cli-babel` configuration (with support for older style with a deprecation). This removes one of the "double duties" that the `babel` config currently takes and _should_ help us traverse the path forward to babel@5 and babel@6 interop. (https://github.com/babel/ember-cli-babel/issues/102)
- [ ] Update ember-cli to **require** `ember-cli-babel` for things to function:
  - [ ] [lib/broccoli/ember-app.js#L1237](https://github.com/ember-cli/ember-cli/blob/dbf034b13cd25cfc401b55019f15d104d4289f09/lib/broccoli/ember-app.js#L1237) - This comment infers something that simply doesn't work anymore. We should likely remove that guard and just throw if we don't detect `ember-cli-babel`.
- [ ] Update `ember-cli-babel` to deprecate using the default of "blacklist modules"
  - [x] Change lib/models/addon.js to compile modules. This is used in ember-cli to ensure that we do a two-pass compilation of babel for addons.  The first is done by the addon itself, and the second is done by lib/broccoli/ember-app.js#L929.
  - [x] Remove second pass module transpilation in ember-cli's [lib/broccoli/ember-app.js](https://github.com/ember-cli/ember-cli/blob/dbf034b13cd25cfc401b55019f15d104d4289f09/lib/broccoli/ember-app.js#L929).
- [x] Update `ember-cli` to use its own specific configuration for modules only babel transpilation instead of utilizing `babel` config option.
  - [ ] [lib/broccoli/ember-app.js#L929](https://github.com/ember-cli/ember-cli/blob/dbf034b13cd25cfc401b55019f15d104d4289f09/lib/broccoli/ember-app.js#L929)
  - [ ] [lib/broccoli/ember-app.js#L1179-L1195](https://github.com/ember-cli/ember-cli/blob/dbf034b13cd25cfc401b55019f15d104d4289f09/lib/broccoli/ember-app.js#L1179-L1195)
- [ ] Update `ember-cli`  to read and pass through `.babelrc` in the project and addon root as the default `babel` options.

**Comment by Turbo87** [comment]
@stefanpenner I updated the task list above

**Comment by knownasilya** [comment]
Looks like `broccoli-babel-transpiler` is holding up `ember-cli-babel` (and `emberjs-build`), which is holding up `ember-cli-htmlbars-inline-precompile`.

**Comment by Turbo87** [comment]
@knownasilya https://github.com/babel/broccoli-babel-transpiler/pull/70#issuecomment-223740887

**Comment by Turbo87** [comment]
tl;dr I wasn't that familiar with the code when I started and wanted to make sure the Babel 6 port actually worked before releasing this as an official `v6.0.0`. Since then I've worked with it a little and haven't found any major issues, so we could release it officially soon.

**Comment by stefanpenner** [comment]
@knownasilya its mostly me dragging my heels on reviewing @Turbo87 wonderful PR's

**Comment by nathanhammond** [comment]
@stefanpenner ping

**Comment by mike-north** [comment]
AFAIK [this is blocking general use, or some use of WeakMap](https://github.com/levanto-financial/ember-lodash/pull/22#issuecomment-240839464) in apps that target browsers which would need a polyfill.

**Comment by stefanpenner** [comment]
@mike-north weak-map comes from core-js, it can be included independently of a babel 6 upgrade.

**Comment by Turbo87** [comment]
Updated the task list above with a set of more concrete steps

**Comment by runspired** [comment]
We will need to port our Globals build infrastructure in ember-data to use babel6 as well.  If this is moving forward sooner than the next 2 weeks I will not have any time to work on doing this, but likely can after that.

cc @pangratz @igorT @fivetanley @bmac

**Comment by rwjblue** [comment]
@runspired - The idea is that each addon specifies its own version.  It is **extremely** unlikely that we can move the entire ecosystem over to babel6 at one specific time.  So we plan to allow addons to control their babel version via their existing ember-cli-babel dep's version.

**Comment by stefanpenner** [comment]
> @runspired - The idea is that each addon specifies its own version. It is extremely unlikely that we can move the entire ecosystem over to babel6 at one specific time. So we plan to allow addons to control their babel version via their existing ember-cli-babel dep's version.

yup, this.

**Comment by mdebbar** [comment]
@stefanpenner @rwjblue do you have an idea when it would be possible to use babel 6? I'm excited to get several perf. improvements for free with babel 6!
(e.g. https://twitter.com/soprano/status/831588549870784514).

**Comment by tschoartschi** [comment]
@stefanpenner @rwjblue @mdebbar I'm also interested in the status of this issue. Are there any low hanging fruits so we can contribute or is everything already on track and only the hard stuff is left?

**Comment by mdebbar** [comment]
@stefanpenner @rwjblue I'm also willing to help if there are clear action items to unblock the upgrade.

**Comment by rwjblue** [comment]
ember-cli is being updated to use babel@6 in https://github.com/ember-cli/ember-cli/pull/6828 (shooting for ember-cli@2.13.0-beta.1 as the first tagged release with this).

ember-cli-babel@6.0.0-beta.2 has been released, and should allow using babel@6 at the app or addon layer even with older ember-cli versions (tested apps and addons using 2.11.1 for example). Please note, that you will likely also have to update ember-cli-htmlbars-inline-precompile to 0.4.0-beta.1 as well.

If folks would like to test either the PR in https://github.com/ember-cli/ember-cli/pull/6828 or existing apps / addons with `ember-cli-babel@6.0.0-beta.2` and `ember-cli-htmlbars-inline-precompile@0.4.0-beta.1` and report back that would be great...

**Comment by rwjblue** [comment]
Obviously, things are still somewhat in flux (and larger breaking changes could happen), but things are looking good for a quickish migration to babel@7 when it is released. I've done some rudimentary tests with babel@7 as well and it should be a much easier migration for us.

**Comment by backspace** [comment]
I tried to apply the new versions @rwjblue mentioned to `travis-web` but we use [`ember-computed-decorators`](https://github.com/rwjblue/ember-computed-decorators) and I couldn’t figure out how to get Babel to accept it, even using [`babel-plugin-transform-decorators-legacy`](https://github.com/loganfsmyth/babel-plugin-transform-decorators-legacy). But I’m looking forward to this!

**Comment by rwjblue** [comment]
I submitted https://github.com/travis-ci/travis-web/pull/983 to show how to do the decorators part. Still looking into what is going on with phantom + babel polyfill...
